packages/repository/src/relations/has-many/has-many-through-repository.factory.ts

```diff
@@ -0,0 +1,219 @@
+import {constrainFilter} from '../../repositories/constraint-utils';
+import {isTypeResolver} from '../../type-resolver';
+import {Getter, HasManyThroughDefinition} from '../relation.types';
+import {
+  DefaultHasManyThroughRepository,
+  HasManyThroughRepository,
+} from './has-many-through.repository';
+import {Filter, Where} from '../../query';
+
+const debug = debugFactory('loopback:repository:has-many-repository-factory');
+
+export type HasManyThroughRepositoryFactory<
+  Target extends Entity,
+  Through extends Entity,
+  ForeignKeyType
+> = (fkValue: ForeignKeyType) => HasManyThroughRepository<Target, Through>;
+
+export interface AdvancedConstraint<T extends Entity> {
+  dataObject: DataObject<T>;
+  filter: Filter<T>;
+  where: Where<T>;
+}
+
+/**
+ * Enforces a constraint on a repository based on a relationship contract
+ * between models. For example, if a Customer model is related to an Order model
+ * via a HasMany relation, then, the relational repository returned by the
+ * factory function would be constrained by a Customer model instance's id(s).
+ *
+ * @param relationMetadata The relation metadata used to describe the
+ * relationship and determine how to apply the constraint.
+ * @param targetRepositoryGetter The repository which represents the target model of a
+ * relation attached to a datasource.
+ * @returns The factory function which accepts a foreign key value to constrain
+ * the given target repository
+ */
+export function createHasManyThroughRepositoryFactory<
+  Target extends Entity,
+  TargetID,
+  Through extends Entity,
+  ThroughID,
+  ForeignKeyType
+>(
+  relationMetadata: HasManyThroughDefinition,
+  targetRepositoryGetter: Getter<EntityCrudRepository<Target, TargetID>>,
+  throughRepositoryGetter: Getter<EntityCrudRepository<Through, ThroughID>>,
+): HasManyThroughRepositoryFactory<Target, Through, ForeignKeyType> {
+  const meta = resolveHasManyThroughMetadata(relationMetadata);
+  debug('Resolved HasMany relation metadata: %o', meta);
+  return function(fkValue?: ForeignKeyType) {
+    async function getAdvancedConstraint(
+      targetInstance: Target,
+    ): Promise<AdvancedConstraint<Target | Through>> {
+      return createAdvancedConstraint<
+        Target,
+        Through,
+        ThroughID,
+        ForeignKeyType
+      >(targetInstance, meta, throughRepositoryGetter, fkValue);
+    }
+    return new DefaultHasManyThroughRepository<
+      Target,
+      TargetID,
+      EntityCrudRepository<Target, TargetID>,
+      Through,
+      ThroughID,
+      EntityCrudRepository<Through, ThroughID>
+    >(targetRepositoryGetter, throughRepositoryGetter, getAdvancedConstraint);
+  };
+}
+
+type HasManyThroughResolvedDefinition = HasManyThroughDefinition & {
+  keyTo: string;
+  targetFkName: string;
+  targetPrimaryKey: string;
+};
+
+async function createAdvancedConstraint<
+  Target extends Entity,
+  Through extends Entity,
+  ThroughID,
+  ForeignKeyType
+>(
```

Comment:
Please specify the return type of this function explicitly.

Reply:
👍

Reply:
https://github.com/strongloop/loopback-next/pull/2359/files/ecbbba2468f702ec3469aa325c368bad7a46c5ae#diff-a35f109a989bf5b2ac1206db18756a0cR99